lib/THC/THCReduceApplyUtils.cuh

```diff
@@ -125,7 +237,7 @@ __device__ T reduceBlockWithNThreadLocalReductions(T *smem,
     local = reduceOp(local, next);
   }
 
-  return reduceBlock<T, ReduceOp>(smem, blockDim.x < numVals ? blockDim.x : numVals, local, reduceOp, init);
+  return reduceBlock<T, ReduceOp>(smem, THCCeilDiv(numVals, N), local, reduceOp, init);
```

Comment:
In this case the `THCCeilDiv` should round it up - if we had 513 values in the above case then the first 257 threads should have valid input values.